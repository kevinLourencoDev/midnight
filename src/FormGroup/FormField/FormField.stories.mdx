import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks';
import { action } from '@storybook/addon-actions';
import FormField from './FormField';

<Meta title='FormField' component={FormField} />

# Demo App

## Properties

<Props of={FormField} />

## Basic usage

<Preview>
	<Story name='text'>
		<FormField
			label='to'
			name='field1'
			type='mail'
			placeholder='Fill me'
			required
		/>
	</Story>
	<Story name='password'>
		<FormField
			label='to'
			name='field1'
			type='password'
			placeholder='&#8226;&#8226;&#8226;&#8226;&#8226;'
			required
		/>
	</Story>
	<Story name='mail'>
		<FormField
			label='to'
			name='field1'
			type='mail'
			placeholder='test@test.com'
			required
		/>
	</Story>
</Preview>
