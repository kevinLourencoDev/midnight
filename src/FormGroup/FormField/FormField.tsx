import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useId } from 'react-id-generator';
import styled from 'styled-components';

import {
	StyledText,
	StyledError,
	StyledInput,
	StyledLabel,
	StyledFromGroup,
} from '../Formstyle';

const validateEmail = (email: string) => {
	var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
	return re.test(String(email).toLowerCase());
};

const FormField = styled(
	({
		name,
		value,
		placeholder,
		hintText,
		type,
		handleChange,
		disabled,
		label,
		required,
	}) => {
		// Generate unique id for htmlFor label/input
		const [htmlId] = useId();

		// isTouched when the input is first focused
		const [errorMessage, setErrorMessage] = useState('');

		const validateChange = (event: any): void => {
			const cleanErrorMessage = () => {
				setErrorMessage('');
			};

			// We make the verification only when we have blur at least one time
			if (required === true) {
				if (event.target.value === '') {
					return setErrorMessage('Champs requis');
				} else {
					cleanErrorMessage();
				}
			}

			if (type === 'mail') {
				if (!validateEmail(event.target.value)) {
					return setErrorMessage('Email invalide');
				} else {
					cleanErrorMessage();
				}
			}
		};

		return (
			<StyledFromGroup>
				<StyledLabel htmlFor={htmlId}>{label}</StyledLabel>
				<StyledInput
					id={htmlId}
					name={name}
					type={type}
					value={value}
					placeholder={placeholder}
					onBlur={() => {
						handleChange ? handleChange(event) : null;
						validateChange(event);
					}}
					disabled={disabled}
					required={required}
				/>
				{hintText && <StyledText>{hintText}</StyledText>}
				{errorMessage && <StyledError>{errorMessage}</StyledError>}
			</StyledFromGroup>
		);
	}
)``;

FormField.propTypes = {
	id: PropTypes.string,
	value: PropTypes.string,
	placeholder: PropTypes.string,
	error: PropTypes.string,
	hintText: PropTypes.string,
	type: PropTypes.string,
	onChange: PropTypes.func,
	disabled: PropTypes.bool,
	label: PropTypes.string,
	required: PropTypes.bool,
};

export default FormField;
