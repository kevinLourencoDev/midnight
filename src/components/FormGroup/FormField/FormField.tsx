import React, { useState, SFC } from 'react';
import { useId } from 'react-id-generator';
import { validateEmail } from '../../../utils/validators';

import {
	StyledText,
	StyledError,
	StyledInput,
	StyledLabel,
	StyledFromGroup,
} from '../formStyle';

export interface FormFieldProps {
	name?: string;
	value?: string;
	placeholder?: string;
	hintText?: string;
	type?: string;
	handleChange?: any;
	disabled?: boolean;
	label?: string;
	required?: boolean;
}

const FormField: SFC<FormFieldProps> = ({
	name,
	value,
	placeholder,
	hintText,
	type = 'text',
	handleChange,
	disabled,
	label,
	required,
}) => {
	// Generate unique id for htmlFor label/input
	const [htmlId] = useId();

	// isTouched when the input is first focused
	const [errorMessage, setErrorMessage] = useState('');

	const validateChange = (event: any): void => {
		const cleanErrorMessage = () => {
			setErrorMessage('');
		};

		// We make the verification only when we have blur at least one time
		if (required === true) {
			if (event.target.value === '') {
				return setErrorMessage('Champs requis');
			} else {
				cleanErrorMessage();
			}
		}

		if (type === 'mail') {
			if (!validateEmail(event.target.value)) {
				return setErrorMessage('Email invalide');
			} else {
				cleanErrorMessage();
			}
		}
	};

	// Define default placeholder
	const defaultPlaceholder = (type: string): string => {
		switch (type) {
			case 'mail':
				return 'Saisissez votre adresse e-mail';
			case 'password':
				return '••••••';
			default:
				return '';
		}
	};

	return (
		<StyledFromGroup>
			<StyledLabel htmlFor={htmlId}>{label}</StyledLabel>
			<StyledInput
				id={htmlId}
				name={name}
				type={type}
				value={value}
				placeholder={placeholder ? placeholder : defaultPlaceholder(type)}
				onBlur={(event) => {
					handleChange ? handleChange(event) : null;
					validateChange(event);
				}}
				disabled={disabled}
				required={required}
			/>
			{hintText && !errorMessage && <StyledText>{hintText}</StyledText>}
			{errorMessage && <StyledError>{errorMessage}</StyledError>}
		</StyledFromGroup>
	);
};

export default FormField;
