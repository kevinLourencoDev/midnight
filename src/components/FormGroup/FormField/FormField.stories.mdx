import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks';
import { action } from '@storybook/addon-actions';
import FormField from './FormField';

<Meta title='FormField' component={FormField} />

# Demo App

## Properties

<Props of={FormField} />

## Basic usage

<Preview>
	<Story name='mail'>
		<FormField label='Votre email' name='mail' type='mail' required />
	</Story>
	<Story name='password'>
		<FormField label='Mot de passe' name='password' type='password' required />
	</Story>
	<Story name='text'>
		<FormField label='Other type possible' name='text' type='text' required />
	</Story>
</Preview>
