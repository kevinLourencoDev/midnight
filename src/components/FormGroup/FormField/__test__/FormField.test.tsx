import React from 'react';
import ReactDom from 'react-dom';
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import '@testing-library/jest-dom/extend-expect';

import FormField from '../FormField';

afterEach(cleanup);

// Create mock of useIf function that break snapshoot test
jest.mock('react-id-generator', () => {
	return {
		useId: jest.fn(() => 'i'),
	};
});

describe('FormField', () => {
	describe('FormField common test', () => {
		it('renders FormField', () => {
			const div: HTMLDivElement = document.createElement('div');

			ReactDom.render(<FormField />, div);
			ReactDom.unmountComponentAtNode(div);
		});

		it('FormField with the attribute required should show error message', () => {
			const wrapper = Enzyme.shallow(<FormField required />);
			wrapper.find('#i').simulate('blur', { target: { value: '' } });

			expect(wrapper.contains('Champs requis')).toBe(true);
		});

		it('FormField with the attribute required should not show error message', () => {
			const wrapper = Enzyme.shallow(<FormField required />);
			wrapper.find('#i').simulate('blur', { target: { value: 'fill' } });

			expect(wrapper.contains('Champs requis')).toBe(false);
		});

		it('FormField without the attribute required should not show error when empty', () => {
			const wrapper = Enzyme.shallow(<FormField />);
			wrapper.find('#i').simulate('blur', { target: { value: '' } });

			expect(wrapper.contains('Champs requis')).toBe(false);
		});

		it('matches snapshot', () => {
			const tree = renderer.create(<FormField />).toJSON();
			expect(tree).toMatchSnapshot();
		});

		it('FormField with placeholder shoud have his own placeholder', () => {
			const wrapper = Enzyme.shallow(
				<FormField placeholder='test placeholder' />
			);

			expect(wrapper.find('#i').prop('placeholder')).toEqual(
				'test placeholder'
			);
		});

		it('FormField without placeholder shoud not have his own placeholder', () => {
			const wrapper = Enzyme.shallow(<FormField />);

			expect(wrapper.find('#i').prop('placeholder')).toEqual('');
		});

		it('Form Fied have an hintText attribute have to show it', () => {
			const wrapper = Enzyme.shallow(<FormField hintText='test hint' />);

			expect(wrapper.contains('test hint')).toBe(true);
		});
	});

	describe('FormField mail test', () => {
		it('renders FormField mail', () => {
			const div: HTMLDivElement = document.createElement('div');

			ReactDom.render(<FormField type='mail' />, div);
			ReactDom.unmountComponentAtNode(div);
		});

		it('Form Fied handleChange must be called when input change', () => {
			const mockHandleChange = jest.fn(() => true);

			const input = Enzyme.shallow(
				<FormField type='mail' handleChange={mockHandleChange} />
			);
			input.find('#i').simulate('blur', { target: { value: 'test@test.com' } });
			expect(mockHandleChange.mock.calls.length).toEqual(1);
		});

		it('Form Fied have to show an error when mail is wrong', () => {
			const wrapper = Enzyme.shallow(<FormField type='mail' />);
			wrapper.find('#i').simulate('blur', { target: { value: 'notamail' } });

			expect(wrapper.contains('Email invalide')).toBe(true);
		});

		it('Form Fied have to not show an error when mail is correct', () => {
			const wrapper = Enzyme.shallow(<FormField type='mail' />);
			wrapper
				.find('#i')
				.simulate('blur', { target: { value: 'mail@mail.com' } });

			expect(wrapper.contains('Email invalide')).toBe(false);
		});

		it('FormField mail without placeholder shoud have a default one', () => {
			const wrapper = Enzyme.shallow(<FormField type='mail' />);

			expect(wrapper.find('#i').prop('placeholder')).toEqual(
				'Saisissez votre adresse e-mail'
			);
		});

		it('FormField password with placeholder shoud have his own placeholder', () => {
			const wrapper = Enzyme.shallow(
				<FormField type='mail' placeholder='test mail enter' />
			);

			expect(wrapper.find('#i').prop('placeholder')).toEqual('test mail enter');
		});

		it('matches snapshot', () => {
			const tree = renderer.create(<FormField type='mail' />).toJSON();
			expect(tree).toMatchSnapshot();
		});
	});

	describe('FormField password test', () => {
		it('renders FormField password', () => {
			const div: HTMLDivElement = document.createElement('div');

			ReactDom.render(<FormField type='password' />, div);
			ReactDom.unmountComponentAtNode(div);
		});

		it('FormField password without placeholder shoud have a default one', () => {
			const wrapper = Enzyme.shallow(<FormField type='password' />);

			expect(wrapper.find('#i').prop('placeholder')).toEqual('••••••');
		});

		it('FormField password with placeholder shoud have his own placeholder', () => {
			const wrapper = Enzyme.shallow(
				<FormField type='password' placeholder='test enter password' />
			);

			expect(wrapper.find('#i').prop('placeholder')).toEqual(
				'test enter password'
			);
		});

		it('matches snapshot', () => {
			const tree = renderer.create(<FormField type='password' />).toJSON();
			expect(tree).toMatchSnapshot();
		});
	});
});
