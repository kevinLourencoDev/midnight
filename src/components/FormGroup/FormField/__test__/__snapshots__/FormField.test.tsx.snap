// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField FormField common test matches snapshot 1`] = `
<div
  className="sc-bdVaJa eYxSpF"
>
  <label
    className="sc-bxivhb bLTrWe"
    htmlFor="i"
  />
  <input
    className="sc-ifAKCX cKjkbQ"
    id="i"
    onBlur={[Function]}
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`FormField FormField mail test matches snapshot 1`] = `
<div
  className="sc-bdVaJa eYxSpF"
>
  <label
    className="sc-bxivhb bLTrWe"
    htmlFor="i"
  />
  <input
    className="sc-ifAKCX cKjkbQ"
    id="i"
    onBlur={[Function]}
    placeholder="Saisissez votre adresse e-mail"
    type="mail"
  />
  
</div>
`;

exports[`FormField FormField password test matches snapshot 1`] = `
<div
  className="sc-bdVaJa eYxSpF"
>
  <label
    className="sc-bxivhb bLTrWe"
    htmlFor="i"
  />
  <input
    className="sc-ifAKCX cKjkbQ"
    id="i"
    onBlur={[Function]}
    placeholder="••••••"
    type="password"
  />
  
</div>
`;
